### PR TITLE
Qualified importing _ctypes

### PR DESCRIPTION
**Pull request by erikvansebille**
To avoid error on `Kernel.__del__`, see also #791

### CODE REVIEW

**Comment by angus-g** [comment]
Sorry to be a bother on this, but this doesn't seem to have fixed the error on my travis build! Any thoughts?

**Comment by angus-g** [comment]
Maybe as an alternative:

```python
import weakref
...
class Kernel(object):
...
    # old __del__:
    def _cleanup(self):
        if self._lib is not None:
            print("cleaning up, ctp:", ctp)
            ctp.FreeLibrary(self._lib._handle) if platform == 'win32' else ctp.dlclose(self._lib._handle)
            del self._lib
            self._lib = None
            if path.isfile(self.lib_file) and self.delete_cfiles:
                [remove(s) for s in [self.src_file, self.lib_file, self.log_file]]

    def compile(self, compiler):
        """ Writes kernel code to file and compiles it."""
        with open(self.src_file, 'w') as f:
            f.write(self.ccode)
        compiler.compile(self.src_file, self.lib_file, self.log_file)
        logger.info("Compiled %s ==> %s" % (self.name, self.lib_file))

        weakref.finalize(self, self._cleanup)
```

From the docs for `weakref.finalize`:
> Return a callable finalizer object which will be called when obj is garbage collected. Unlike an ordinary weak reference, a finalizer will always survive until the reference object is collected, greatly simplifying lifecycle management.

> A finalizer will never invoke its callback during the later part of the interpreter shutdown when module globals are liable to have been replaced by None.

But perhaps the finaliser should be registered on the `Kernel` object, and it can be marked dead with `finalizer.detach()` if `Kernel.remove_lib()` is called?

Thoughts?

**Comment by erikvansebille** [comment]
Does that solution with `finalizer` work on your Travis? If so, then it might be a good solution yes!

Note that we could also use a similar approach in `ParticleFile.__del__`, which can experience the same problem with unlinked objects that cannot be deleted

**Comment by angus-g** [comment]
Just tested, and yes, using the finalizer seems to clean up correctly (no files left in `/tmp/parcels-1000/...`) and not throw errors, both on my test container and Travis. I think it might be a more robust solution for `ParticleFile` too: I think if you explicitly call the finalizer object, it will be dead and won't be called again, so it implicitly has the `convert_at_end` logic.

I can create a new PR for my `Kernel` changes?

**Comment by erikvansebille** [comment]
> Just tested, and yes, using the finalizer seems to clean up correctly (no files left in `/tmp/parcels-1000/...`) and not throw errors, both on my test container and Travis. I think it might be a more robust solution for `ParticleFile` too: I think if you explicitly call the finalizer object, it will be dead and won't be called again, so it implicitly has the `convert_at_end` logic.

OK, great. That's a good solution then

> I can create a new PR for my `Kernel` changes?

Yes please

**Comment by erikvansebille** [comment]
Fixed in #794
